File: SDK_Examples/C++/ChangePins.cpp
#include "eidlib.h"
#include "eidlibException.h"
#include <iostream>

using namespace eIDMW;

int main(int argc, char **argv) {

    if (argc != 2) 
    {
        std::cout << "Usage: ChangePins [pin]" << std::endl; 
        std::cout << "Pin:" << std::endl; 
        std::cout << "\t-auth\t\tChange Authentication Pin." << std::endl; 
        std::cout << "\t-sign\t\tChange Signature Pin." << std::endl; 
        std::cout << "\t-addr\t\tChange Address Pin." << std::endl; 
        return -1;
    }
    std::string type_flag = argv[1];
    int ret = 0;

    try 
    {
        //Initializes SDK (must always be called in the beginning of the program)
        PTEID_InitSDK();

        //Gets the set of readers connected to the system
        PTEID_ReaderSet& readerSet = PTEID_ReaderSet::instance();
        
        //Gets a reader connected to the system (useful if you only have one)
        //Alternatively you can iterate through the readers using getReaderByNum(int index) instead of getReader()
        PTEID_ReaderContext& readerContext = readerSet.getReader();
        
        //Gets the EIDCard and EId objects (with the cards information)
        PTEID_EIDCard& eidCard = readerSet.getReader().getEIDCard();

        //The number of PIN verification tries that the user has (updated with each call to verifyPin or changePin)
        unsigned long triesLeft;

        //Set of the card PINs
        PTEID_Pins& pins = eidCard.getPins();
        unsigned long pinRef = 0;
        
        //Get the specific PIN we want and proceeds to change it
        //ADDR_PIN - Address Pin
        //AUTH_PIN - Authentication Pin
        //SIGN_PIN - Signature Pin
        if (!type_flag.compare("-addr"))
        {
            pinRef = PTEID_Pin::ADDR_PIN;
        } 
        else if (!type_flag.compare("-auth")) 
        {
            pinRef = PTEID_Pin::AUTH_PIN;
        }
        else if (!type_flag.compare("-sign")) 
        {
            pinRef = PTEID_Pin::SIGN_PIN;
        }
        else 
        {
            std::cout << "Pin type doesn't exist!" << std::endl;
            return -1;
        }

        PTEID_Pin& pin = pins.getPinByPinRef(pinRef);

        bool bResult = pin.changePin("","", triesLeft, NULL);

        std::cout << "PIN sucessfully changed?: " << bResult << " Tries left: " << triesLeft << std::endl;

    }
    catch (PTEID_ExNoReader &e) 
    {
        std::cout << "No readers found!" << std::endl;
        ret = -2;
    }
    catch (PTEID_ExNoCardPresent &e) 
    {
        std::cout << "No card found in the reader!" << std::endl;
        ret = -3;
    }
    catch (PTEID_Exception &e) 
    {
        std::cout << "Caught exception in some SDK method. Error: " << e.GetMessage() << std::endl;
        ret = -4;
    }

    //Releases SDK (must always be called at the end of the program)
    PTEID_ReleaseSDK();
    return ret;
}

File: SDK_Examples/C#/Examples/ChangePins.cs
using System;
using System.IO;
using pt.portugal.eid;

namespace Examples
{
    class ChangePins
    {
        //Main attributes needed for SDK functionalities
        PTEID_ReaderSet readerSet = null;
        PTEID_ReaderContext readerContext = null;
        PTEID_EIDCard eidCard = null;
        PTEID_EId eid = null;
        PTEID_CardType cardType = PTEID_CardType.PTEID_CARDTYPE_UNKNOWN;
        PTEID_CardContactInterface contactInterface = PTEID_CardContactInterface.PTEID_CARD_CONTACTEMPTY;



        /*
         * Initializes the SDK and sets main variables
         */
        public void Initiate()
        {
            //Must always be called in the beginning of the program
            PTEID_ReaderSet.initSDK();

            //Enable test mode if you're using a test card
            //PTEID_Config.SetTestMode(true);

            //Gets the set of connected readers, if there is any inserted
            readerSet = PTEID_ReaderSet.instance();

            //Gets the first reader
            //When multiple readers are connected, you should iterate through the various indexes with the methods getReaderName and getReaderByName
            readerContext = readerSet.getReader();

            //Gets the Card Contact Interface and type

            if (readerContext.isCardPresent())
            {
                contactInterface = readerContext.getCardContactInterface();
                cardType = readerContext.getCardType();
                Console.WriteLine("Contact Interface:" + (contactInterface == PTEID_CardContactInterface.PTEID_CARD_CONTACTLESS ? "CONTACTLESS" : "CONTACT"));
            }

            //Gets the card instance
            eidCard = readerContext.getEIDCard();

            //If the contactInterface is contactless and the card supports contactless then authenticate with PACE
            if (contactInterface == PTEID_CardContactInterface.PTEID_CARD_CONTACTLESS && cardType == PTEID_CardType.PTEID_CARDTYPE_IAS5)
            {
                Console.WriteLine("Insert the Card access number (CAN) for the card in use: ");
                string can_str = Console.ReadLine();
                uint can_size = (uint)can_str.Length;
                eidCard.initPaceAuthentication(can_str, can_size, PTEID_CardPaceSecretType.PTEID_CARD_SECRET_CAN);
            }
            eid = eidCard.getID();
        }

        /*
         * Releases the SDK (must always be done at the end of the program)
         */
        public void Release()
        {
            try
            {
                PTEID_ReaderSet.releaseSDK();
            }
            catch (PTEID_Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
        }

        /*
         * Changes the pin code
         */
        public void ChangePin(String pin_type)
        {

            //Gets authentication pin
            PTEID_Pins pins = eidCard.getPins();
            //The number of tries that the user has (updated with each call to verifyPin)
            uint triesLeft = uint.MaxValue;
            PTEID_Pin pin;

            //Gets the specific PIN we want
            //ADDR_PIN - Address Pin
            //AUTH_PIN - Authentication Pin
            //SIGN_PIN - Signature Pin
            if (pin_type.Equals("-addr"))
            {
                pin = pins.getPinByPinRef(PTEID_Pin.ADDR_PIN);
            }
            else if (pin_type.Equals("-auth"))
            {
                pin = pins.getPinByPinRef(PTEID_Pin.AUTH_PIN);
            }
            else if (pin_type.Equals("-sign"))
            {
                pin = pins.getPinByPinRef(PTEID_Pin.SIGN_PIN);
            }
            else
            {
                throw new Exception("Pin type doesn't exist");
            }

            //Changes the pin
            pin.changePin("", "", ref triesLeft, pin.getLabel(), false);
            Console.WriteLine("Pin changed successfully.");
        }

        public void start(string[] args)
        {
            try
            {
                Initiate();
                ChangePin(args[0]);
            }
            catch (PTEID_ExNoReader)
            {
                Console.WriteLine("No reader found.");
            }
            catch (PTEID_ExNoCardPresent)
            {
                Console.WriteLine("No card inserted.");
            }
            catch (PTEID_Exception ex)
            {
                Console.WriteLine(ex.GetMessage());
            }
            finally
            {
                Release();
                Console.ReadLine();
            }
        }

        static void Main(string[] args)
        {
            new ChangePins().start(args);
        }
    }
}
